src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -1820,26 +1835,71 @@ private static void AddInferredAcceptsMetadata(RequestDelegateFactoryContext fac
         factoryContext.EndpointBuilder.Metadata.Add(new AcceptsMetadata(type, factoryContext.AllowEmptyRequestBody, contentTypes));
     }
 
-    private static Expression BindParameterFromFormFiles(
+    private static Expression BindParameterFromFormCollection(
         ParameterInfo parameter,
         RequestDelegateFactoryContext factoryContext)
     {
-        if (factoryContext.FirstFormRequestBodyParameter is null)
+        // Do not duplicate the metadata if there are multiple form parameters
+        if (!factoryContext.ReadForm)
         {
-            factoryContext.FirstFormRequestBodyParameter = parameter;
+            AddInferredAcceptsMetadata(factoryContext, parameter.ParameterType, FormContentType);
         }
 
+        factoryContext.FirstFormRequestBodyParameter ??= parameter;
         factoryContext.TrackedParameters.Add(parameter.Name!, RequestDelegateFactoryConstants.FormFileParameter);
+        factoryContext.ReadForm = true;
+
+        return BindParameterFromExpression(
+            parameter,
+            FormExpr,
+            factoryContext,
+            "body");
+    }
+
+    private static Expression BindParameterFromFormItem(
+        ParameterInfo parameter,
+        string key,
+        RequestDelegateFactoryContext factoryContext)
+    {
+        var valueExpression = GetValueFromProperty(FormExpr, FormIndexerProperty, key, GetExpressionType(parameter.ParameterType));
 
         // Do not duplicate the metadata if there are multiple form parameters
         if (!factoryContext.ReadForm)
+        {
+            AddInferredAcceptsMetadata(factoryContext, parameter.ParameterType, FormContentType);
+        }
+
+        factoryContext.FirstFormRequestBodyParameter ??= parameter;
+        factoryContext.TrackedParameters.Add(key, RequestDelegateFactoryConstants.FormAttribute);
+        factoryContext.ReadForm = true;
+
+        return BindParameterFromValue(
+            parameter,
+            valueExpression,
+            factoryContext,
+            "form");
+    }
+
+    private static Expression BindParameterFromFormFiles(
+        ParameterInfo parameter,
+        RequestDelegateFactoryContext factoryContext)
+    {
+        // Do not duplicate the metadata if there are multiple form file parameters
+        if (!factoryContext.ReadFormFile)
```

Comment:
I'm not very familiar with how forms work, but is there any way to have this code path run and register just "multipart/form-data", then have one of the other paths run which would normally add "application/x-www-form-urlencoded" as well, but now they won't because we set `ReadForm` below.

Reply:
I am not sure if I get your question correctly but, while might be possible send file using `application/x-www-form-urlencoded` (i am not very familiar 🤷‍♂️ ) . The `FormFeature` only read files from a `multipart/form-data` request.

https://github.com/dotnet/aspnetcore/blob/8dc432556c937465edb2cdff4ab2ad7cbb16bd92/src/Http/Http/src/Features/FormFeature.cs#L171

My approach here is to keep not adding duplicated metadata, however, since we don't know the parameter ordering and we process them sequentially, in case we found a `IFormFile` or `IFormFileCollection` an additional is added, for the first parameter only, with `multipart/form-data` and since the last metadata is the most relevant it should basically this one been used, make sense?